Implement steady state solver using LinearSolve.jl and eigen

LorenzoFioroni opened the pull request:
The goal of this pull request is to make a step forward in the direction of fixing issue #81

## General description

In issue #81 it has been asked to implement, along with the `SteadyStateDirectSolver` already present within the library, an iterative solver using LinearSolve.jl and an eigen solver using the already implemented `eigsolve` function.

I aimed to make the code as agnostic as possible and compatible with both dense and sparse matrices on CPU and GPU. I haven't been able to make some combinations of method-device work yet, but I think this might be a good starting point.

## Summary of the changes

I introduced, along with the already implemented `SteadyStateDirectSolver`, two new structs: `SteadyStateEigenSolver` and `SteadyStateLinearSolver`. While the first two don't require any argument when created, the latter allows specifying optional parameters (see below).

### Direct solver

No changes have been made on this side, and the code remains as originally implemented in the library. 

From a few tests, this method seems to work on sparse and dense CPU matrices and dense GPU ones, but not on sparse GPU matrices. It seems that the left-division operation on GPU matrices results in an indexing of the array, which is disabled by `CUDA.allowscalar(false)` for efficiency reasons.

### Linear solver

The `SteadyStateLinearSolver` accepts three optional keyword arguments, allowing the specification of the solving algorithm from the LinearSolve.jl package, and the left- and right- preconditioners. Since the preconditioners are applied to a matrix `L_tmp` analogous to that defined within the direct solver (see [here](https://github.com/qutip/QuantumToolbox.jl/blob/9786ddf796fa29a23766a5cba0ed6397ad34ead7/src/steadystate.jl#L45)), they are defined as functions. For example, `SteadyStateLinearSolver(alg=KrylovJL_GMRES(), Pl=L_tmp -> ilu(L_tmp, τ = 0.01))` creates a `SteadyStateLinearSolver` which uses the `KrylovJL_GMRES` algorithm defined in LinearSolve.jl and performs a preconditioning with the `ilu` method from the IncompleteLU.jl package. If no algorithm is specified, LinearSolve.jl uses a direct method analogous to our DirectSolver. Moreover, additional parameters can be given to the `solve` function of LinearSolve.jl by passing them as keyword arguments to the call to `steadystate`.

When used as a direct solver, it seems that this method only fails for sparse GPU arrays. I think the reason might be the call to `SparseMatrixCSC()` in [LinearSolve.jl](https://github.com/SciML/LinearSolve.jl/blob/270b56d864e91c2873d27c71b78115a04b824e23/src/factorization.jl#L1359C1-L1360C1), which is not implemented when the parameters are GPU arrays

When used with an iterative method (e.g., `KrylovJL_GMRES`), this solver works for all the combinations of sparse-dense matrices and both on CPU and GPU. Yet, the application to sparse GPU matrices is very slow compared to that on CPU, because no preconditioner is applied. From the [preconditioners page](https://docs.sciml.ai/LinearSolve/stable/basics/Preconditioners/) of LinearSolve.jl's docs, the only GPU-ready preconditioners are defined in KrylovPreconditioners.jl. However, taking for example the `kp_ilu0` method, this seems to be only defined for real CSC matrices ([see here](https://github.com/JuliaSmoothOptimizers/KrylovPreconditioners.jl/blob/09af6873d27e425f8a757e4c08ae5faa0fd9b664/ext/CUDA/ilu0.jl#L97)) while it would be implemented for complex ones as well if they were CSR ([here](https://github.com/JuliaSmoothOptimizers/KrylovPreconditioners.jl/blob/09af6873d27e425f8a757e4c08ae5faa0fd9b664/ext/CUDA/ilu0.jl#L83)).

### EigenSolver

The EigenSolver is nothing but an interface to the already implemented `eigsolve` function. Any keyword arguments specified in the call to the `steadystate` function will be passed to `eigsolve`.

This method seems to work only for CPU matrices, as I consistently get a _scalar indexing_ error from the `arnoldi_init` function for GPU matrices.

---

I hope this can be a good starting point to improve the support of different solving methods within the `steadystate` function. Please let me know what you think and if there is anything I could improve already. Thank you!

albertomercurio commented:
Great! I think that the most important thing is to make the code array-agnostic, that's what you have already done.

The fact that some of the operations are not yet supported on GPU, doesn't mean that they won't be fixed in the future. For example, we just need to wait for **KrylovPreconditioners.jl** supports the preconditioner for `CuSparseMatrixCSC`.

A related question is: what happens if you consider a `CuSparseMatrixCSR` with preconditioner?

BTW, don't forget to add some runtests, in order to check that everything works well.

albertomercurio commented:
Moreover, do you see any performance downgrade using the direct solver of **LinearSolve.jl** with respect to Julia's default?

LorenzoFioroni commented:
Thanks for the feedback!

I've added a test for each method, comparing the steady-state density matrix to that obtained by integrating the Lindblad master equation. The comparison is done via the recently added `tracedist` function. 

Indeed, using the direct method through LinearSolve.jl results in a little downgrade of the solver's performance. I've switched back to using the original direct solver as the default choice.

Finally, I formatted the code, which I had forgotten to do before - sorry about that!

albertomercurio commented:
Great. You have to format also the test file. Then everything seems to be fine.
